### PR TITLE
Balances

### PR DESCRIPTION
**Pull request by mortonjt**
Adding in a new semantic type to represent balances.  This will make it much easier for different machine learning techniques and statistical analyses to operate directly on balances.

Note that this will be an important component for the upcoming gneiss release.

### CODE REVIEW

**Comment by ebolyen** [comment]
For my own edification, what constraints does this new variant-type imply?

**Comment by mortonjt** [comment]
This balance type doesn't really have constraints.  Since a single balance refers a log ratio, that log ratio can stretch anywhere from negative infinity or positive infinity.  

The main reason I'm hoping to have this in the `q2-composition` package is to make it easier to pass inputs from various q2 packages easier.  For one, there are a few different R packages that perform similar calculations with balances.  In addition, it would be great if it were possible to directly pass balances into other techniques, such as random forests, logistic regression, PCA, ...

**Comment by mortonjt** [comment]
Also, the Balance type can only be generated from a Composition type.

And this can only be done using the ilr transform. We'll have a conversion
function in gneiss shortly after this is merged.

On Mar 30, 2017 11:25 AM, "Evan Bolyen" <notifications@github.com> wrote:

For my own edification, what constraints does this new variant-type imply?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/qiime2/q2-composition/pull/34#issuecomment-290482020>,
or mute the thread

**Comment by mortonjt** [comment]
I briefly discussed this with @ebolyen .

The main concern with this PR is that this type `Balance` is not generic enough, and may not differentiate between the different types of balances.

To give a little context, there are two types of balances.  There are the balances that are computed using the ilr transform, which is used by gneiss and phylofactor.  There is another type of balance which is used in [philr](http://bioconductor.org/packages/release/bioc/html/philr.html) that can be calculated using the weighted ilr transform, which incorporates branch length.

The standard ilr transform will cover the vast majority of the use cases involving balances.  If we want to add philr to the standard q2 workflow, we can add a `WeightedBalance` type.  Does this sound like a reasonable plan forward @ebolyen @gregcaporaso ?

**Comment by gregcaporaso** [comment]
This sounds reasonable to me, but I'm interested in what @ebolyen thinks as you and he discussed it.

**Comment by ebolyen** [comment]
I think this makes sense, but I don't really understand compositional statistics.

My main concern was whether there would be other transformations which could also be considered a balance. This could be a problem if you used two different transforms, producing two `FeatureTable[Balance]` artifacts and then tried to merge them (which doesn't seem like a valid operation to do on compositional data anyhow).

After reading a bit and looking at some of slides @mortonjt's shared a while back, it sounds like the balance is what is used to create the basis for the ILR transform, so this term doesn't really apply to other transforms, which don't use the orthonormal basis (how wrong am I here?).

In which case, I think these type names sound great, and `WeightedBalance` also makes sense for `philr` I think.

**Comment by mortonjt** [comment]
I've updated the type definition to fit in with the new updates.

If there are no objections - do you guys mind if we could merge this in?  This would make the refactoring process in Gneiss much easier.  I'm currently running into import issues because of this new type.

File: q2_composition/plugin_setup.py [hunk 1]
```diff
@@ -40,6 +40,18 @@
...
+
+plugin.register_semantic_type_to_format(
+    FeatureTable[Balance],
+    artifact_format=BIOMV210DirFmt
+)
+
+plugin.register_semantic_type_to_format(
+    FeatureTable[Balance],
+    artifact_format=BIOMV100DirFmt
+)
```
Comment (ebolyen) [inline q2_composition/plugin_setup.py]: You should drop this registration (and the similar one for `Composition` to `BIOMV100DirFmt`). It doesn't actually do anything, it's still left-over from `q2-types` for a situation where we abused the framework's lookup for a dynamic unit-test (none of those were *great* ideas, but it worked well enough at the time).

Comment (mortonjt) [inline q2_composition/plugin_setup.py]: Thanks @ebolyen !

**Comment by mortonjt** [comment]
Thanks!

**Pull request merged**